File: examples/serialization/src/main/java/japi/ConditionalExample.java
/*
 * Copyright 2015 - 2016 Red Bull Media House GmbH <http://www.redbullmediahouse.com> - all rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package japi;

import static japi.DocUtils.append;
import japi.ViewExample.GetAppendCountCommandReply;

//#conditional-requests
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.japi.pf.ReceiveBuilder;
import akka.pattern.Patterns;
import akka.util.Timeout;
import com.rbmhtechnology.eventuate.*;
import scala.concurrent.ExecutionContextExecutor;

import java.util.Collection;
import java.util.Collections;
import java.util.Optional;
import java.util.concurrent.TimeUnit;

import static scala.compat.java8.JFunction.func;
import static scala.compat.java8.JFunction.proc;
//#

public class ConditionalExample {
  //#conditional-requests

  class ExampleActor extends AbstractEventsourcedActor {
    private final String id;
    private Collection<String> currentState = Collections.emptyList();
    private Messages msgs = new Messages();

    public ExampleActor(String id, ActorRef eventLog) {
      super(id, eventLog);
      this.id = id;

      setOnCommand(ReceiveBuilder
        .match(Messages.AppendCommand.class, cmd -> persist(msgs.new AppendedEvent(cmd.entry), ResultHandler.onSuccess(
          evt -> sender().tell(msgs.new AppendSuccessWithTimestampCommandReply(evt.entry, lastVectorTimestamp()), self())
        )))
        // ...
        .build());

      setOnEvent(ReceiveBuilder
        .match(Messages.AppendedEvent.class, evt -> currentState = append(currentState, evt.entry))
        .build());
    }

    @Override
    public Optional<String> getAggregateId() {
      return Optional.of(id);
    }
  }

  // Eventsourced-View
  class ExampleView extends AbstractEventsourcedView {
    // AbstractEventsourcedView has ConditionalRequests mixed-in by default

    public ExampleView(String id, ActorRef eventLog) {
      super(id, eventLog);

      // ...
    }
  }
  //#
  public static void main(String[] args) {
    final ActorSystem system = ActorSystem.create("");
    final ActorRef eventLog = null;
    final ExecutionContextExecutor dispatcher = system.dispatcher();

    //#conditional-requests

    ConditionalExample ce = new ConditionalExample();
    Messages msgs = new Messages();
    final ActorRef ea = system.actorOf(Props.create(ExampleActor.class, () -> ce.new ExampleActor("ea", eventLog)));
    final ActorRef ev = system.actorOf(Props.create(ExampleView.class, () -> ce.new ExampleView("ev", eventLog)));

    final Timeout timeout = Timeout.apply(5, TimeUnit.SECONDS);

    Patterns.ask(ea, msgs.new AppendCommand("a"), timeout)
      .flatMap(func(m ->
          Patterns.ask(ev,
              new ConditionalRequest(((Messages.AppendSuccessWithTimestampCommandReply) m).updateTimestamp, ce.new GetAppendCount()), timeout)
          )
        , dispatcher)
      .onComplete(proc(result -> {
        if (result.isSuccess()) {
          System.out.println("append count = " + ((GetAppendCountCommandReply) result.get()).count);
        }
      }), dispatcher);

    //#
  }

  class GetAppendCount {}
}

File: examples/user-guide/src/main/scala/sapi/ConditionalExample.scala
/*
 * Copyright 2015 - 2017 Red Bull Media House GmbH <http://www.redbullmediahouse.com> and Mike Slinn - all rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package sapi

object ConditionalExample extends App {
  import akka.actor._
  //#conditional-requests
  import akka.pattern.ask
  import akka.util.Timeout
  import com.rbmhtechnology.eventuate._
  import EventsourcedViews._
  import scala.concurrent.duration._
  import scala.util._

  //#
  import com.rbmhtechnology.eventuate.log.leveldb.LeveldbEventLog

  implicit val system: ActorSystem = ActorSystem(ReplicationConnection.DefaultRemoteSystemName)

  // This will create a directory called `target/log-qt-2/` to contain the log files
  val eventLog: ActorRef = system.actorOf(LeveldbEventLog.props("qt-2"))

  //#conditional-requests
  class ExampleActor(
    override val id: String,
    override val eventLog: ActorRef
  ) extends EventsourcedActor {
    private var currentState: Vector[String] = Vector.empty
    override val aggregateId = Some(id)

    override def onCommand: PartialFunction[Any, Unit] = {
      case AppendCommand(entry) => persist(AppendedEvent(entry)) {
        case Success(_) =>
          sender() ! AppendSuccessWithTimestampCommandReply(entry, lastVectorTimestamp)

        case Failure(_) =>
          // ...
      }
      // ...
    }

    override def onEvent: PartialFunction[Any, Unit] = {
      case AppendedEvent(entry) => currentState = currentState :+ entry
    }
  }

  class ExampleView(override val id: String, override val eventLog: ActorRef)
    extends EventsourcedView with ConditionalRequests {
    // ...
  //#
    private var appendCount: Long = 0L
    private var resolveCount: Long = 0L

    override def onCommand: PartialFunction[Any, Unit] = {
      case GetAppendCount => sender() ! GetAppendCountReply(appendCount)
      case GetResolveCount => sender() ! GetResolveCountReply(resolveCount)
    }

    override def onEvent: PartialFunction[Any, Unit] = {
      case _: AppendedEvent => appendCount += 1L
      case _: Resolved => resolveCount += 1L
    }
  //#conditional-requests
  }

  val ea: ActorRef = system.actorOf(Props(new ExampleActor("ea", eventLog)))
  val ev: ActorRef = system.actorOf(Props(new ExampleView("ev", eventLog)))

  import system.dispatcher
  implicit val timeout = Timeout(5.seconds)

  for {
    AppendSuccessWithTimestampCommandReply(_, timestamp) <- ea ? AppendCommand("a")
    GetAppendCountReply(count)  <- ev ? ConditionalRequest(timestamp, GetAppendCount)
  } println(s"append count = $count")
  //#

  Util.pauseThenStop()
}
